File: contest_practice_solution.cpp
#include <bits/stdc++.h>
using namespace std;


int dfs(int node, vector<vector<int>>& adj_list, vector<bool>& visited) {
    visited[node] = true;
    bool is_leaf=true;
    int total = 0;
    for (int child : adj_list[node]) {
        if (!visited[child]) {
           is_leaf=false;
           total+= dfs(child, adj_list, visited);
        }
    }
    if(is_leaf){
        return 1;
    }
    return total;
}

int main() {
    int N;
    cin >> N;
    
    vector<vector<int>> adj_list(N + 1);
    vector<bool> visited(N + 1, false);

    for (int i = 1; i<N; i++) {
        int a, b;
        cin >> a >> b;
        adj_list[a].push_back(b);
    }
    
    
  
    int sum = dfs(1, adj_list, visited);
    cout<<sum<<"\n";
    
    return 0;
}

File: contest_practice_solution.py
def dfs(node, adj_list, visited, leaf_sum):
    visited[node] = True
    is_leaf = True
    for neighbor in adj_list[node]:
        if not visited[neighbor]:
            is_leaf = False
            leaf_sum = dfs(neighbor, adj_list, visited, leaf_sum)
    if is_leaf:
        leaf_sum += 1
    return leaf_sum


n, x = map(int, input().split())

adj_list = [[] for _ in range(n+1)]


for i in range(n-1):
    u, v = map(int, input().split())
    adj_list[u].append(v)
    adj_list[v].append(u)

visited = [False] * (n+1)
leaf_sum = dfs(1, adj_list, visited, 0)
print(leaf_sum)
